README.md
[![PyPI version](https://badge.fury.io/py/seq-to-first-iso.svg)](https://badge.fury.io/py/seq-to-first-iso)
[![Conda](https://img.shields.io/conda/v/bioconda/seq-to-first-iso.svg?color=bright-green)](https://anaconda.org/bioconda/seq-to-first-iso)
[![Build Status](https://travis-ci.org/pierrepo/seq-to-first-iso.svg?branch=master)](https://travis-ci.org/pierrepo/seq-to-first-iso)
[![Documentation Status](https://readthedocs.org/projects/seq-to-first-iso/badge/?version=latest)](https://seq-to-first-iso.readthedocs.io/en/latest/?badge=latest)

# Seq-to-first-iso

> Compute first two isotopologues intensity from peptide sequence

Seq-to-first-iso computes isotopologues M0 and M1 of peptides with a 99.99 % 12C enrichment for quantification by [SLIM-labeling](https://www.mcponline.org/content/16/11/2017).  
It simulates auxotrophies by differentiating labelled and unlabelled amino acids.

The documentation can be found [here](https://seq-to-first-iso.readthedocs.io/en/latest/).  
Try the demo with Binder: [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/pierrepo/seq-to-first-iso/master)

## Installation

### With pip

```
$ pip install seq-to-first-iso
```

### With conda

```
$ conda install seq-to-first-iso -c bioconda
```

### Developer mode


Install [conda](https://conda.io/projects/conda/en/latest/user-guide/install/index.html)

Clone repo:
```
$ git clone https://github.com/pierrepo/seq-to-first-iso
$ cd seq-to-first-iso
```

Create conda environment:
```
$ conda env create -f environment.yml
```

Remark: for a fully reproducible environment, you could also use:
```
$ conda env create -f environment.lock.yml
```


Activate conda environment:
```
$ conda activate seq-to-first-iso
```

Install local package:
```
$ pip install -e .
```

## Usage

The script takes a file with one sequence of amino acids per line and returns a tsv of the file with columns:

|sequence|mass|formula|formula_X| M0_NC | M1_NC | M0_12C | M1_12C |
|--------|----|-------|---------|-------|-------|--------|--------|

Once installed, the script can be called with:

```shell
$ seq-to-first-iso filename [-o output_name] [-n amino_acids...]
```
Optional arguments are in square brackets  
This will create _filename_stfi.tsv_ if filename is a correct file

**0.3.0 :** The input file can have annotations separated by a tabulation before the sequences  
**0.4.0 :** Support for [X!Tandem](https://www.thegpm.org/tandem/) Post-Translational Modifications added

### Options

- `-h, --help`:  
Provide a help page

- `-v, --version`:  
Provide the version

- `-o, --output`:  
Change the name of the output file

- `-n, --non-labelled-aa`:  
Take 1 or more amino acid separated by a comma


### Examples

- You can provide a list of amino acids which will keep default isotopic abundance:

Supposing [peptides.txt](https://github.com/pierrepo/seq-to-first-iso/blob/master/notebooks/peptides.txt) :

```
YAQEISR
VGFPVLSVKEHK
LAMVIIKEFVDDLK
```

The command
```shell
$ seq-to-first-iso peptides.txt -n V,W
```
will create [peptides_stfi.tsv](https://github.com/pierrepo/seq-to-first-iso/blob/master/notebooks/peptides_stfi.tsv) :

|sequence| mass| formula|formula_X| M0_NC| M1_NC| M0_12C| M1_12C|
|--------|-----|--------|---------|------|------|-------|-------|
YAQEISR| 865.42938099921| C37H59O13N11| C37H59O13N11| 0.6206414140575179|	0.280870823368276| 0.9206561231798033| 0.05161907174495234|
VGFPVLSVKEHK| 1338.7659712609| C63H102O16N16| C48H102O16N16X15|  0.4550358985377136| 0.34506032928190855| 0.7589558393662944| 0.18515489894512063|
LAMVIIKEFVDDLK| 1632.91606619252| C76H128O21N16S1| C66H128O21N16S1X10| 0.36994021481230627| 0.3373188347614264| 0.7475090558698947| 0.15292723586285323|

Where, in 12C enrichment conditions, the isotopologue intensity M0_12C and M1_12C are computed with unlabelled Valine and Tryptophan (V and W have default isotopic abundance)


- You can change the name of the output file:

```shell
$ seq-to-first-iso peptides.txt -o sequence
```
will create a file named *sequence.tsv*


## Credits

- **Binder**
  - Jupyter et al., "Binder 2.0 - Reproducible, Interactive, Sharable Environments for Science at Scale." Proceedings of the 17th Python in Science Conference. 2018. 10.25080/Majora-4af1f417-011

- **Bioconda**:
  - Grüning, Björn, Ryan Dale, Andreas Sjödin, Brad A. Chapman, Jillian Rowe, Christopher H. Tomkins-Tinch, Renan Valieris, the Bioconda Team, and Johannes Köster. 2018. “Bioconda: Sustainable and Comprehensive Software Distribution for the Life Sciences”. *Nature Methods*, 2018 doi:10.1038/s41592-018-0046-7.

- **MIDAs**:
  - Alves G, Ogurtsov AY, Yu YK (2014) Molecular Isotopic Distribution Analysis (MIDAs) with adjustable mass accuracy. *J Am Soc Mass Spectrom*, 25: 57-70. DOI: 10.1007/s13361-013-0733-7

- **Pyteomics**:
  - Goloborodko, A.A.; Levitsky, L.I.; Ivanov, M.V.; and Gorshkov, M.V. (2013) “Pyteomics - a Python Framework for Exploratory Data Analysis and Rapid Software Prototyping in Proteomics”, *Journal of The American Society for Mass Spectrometry*, 24(2), 301–304. DOI: 10.1007/s13361-012-0516-6

  - Levitsky, L.I.; Klein, J.; Ivanov, M.V.; and Gorshkov, M.V. (2018) “Pyteomics 4.0: five years of development of a Python proteomics framework”, *Journal of Proteome Research*. DOI: 10.1021/acs.jproteome.8b00717

- **SLIM-labeling**:
  - Léger T, Garcia C, Collomb L, Camadro JM. A Simple Light Isotope Metabolic Labeling (SLIM-labeling) Strategy: A Powerful Tool to Address the Dynamics of Proteome Variations In Vivo. *Mol Cell Proteomics*. 2017;16(11):2017–2031. doi:10.1074/mcp.M117.066936


CHANGELOG.md
# Changelog

**Dev**

**1.1.0** (2019-12-19)
- Add export_to_knime() function

**1.0.0** (2019-12-05)
- Simplify interface
- Take .tsv as input file
- Take peptide sequence and charge as input data (from input file)
- Check input file and input dataframe
- Comply with PEP 8 and PEP 257
- Update API and CLI demo notebooks
  
**0.5.1** (2019-07-10)

*Fixed*
- seq_to_xcomp() can now correctly take a pyteomics.mass.Composition object as the second parameter

**0.5.0** (2019-07-03)

*Added*
- Flag for version number in CLI
- Jupyter notebook with Binder environment for demonstrations
- Documentation on Read the Docs
- Conda availability

*Changed*
- **Breaking change** : changed seq_to_midas() to seq_to_xcomp()
- **Breaking change** : changed seq_to_tsv() to seq_to_df()

**0.4.3** (2019-06-26)

*Fixed*
- Fix requirements not being installed with `pip install`

**0.4.2** (2019-06-25)

*Fixed*
- Fix *setup.cfg*'s installation requirements

**0.4.1** (2019-06-24)

- Extend numpydoc style to all functions in *seq_to_first_iso.py*

**0.4.0** (2019-06-21)

*Changed*
- Add support for Xtandem Parsing
  - **Breaking change:** sequence_parser() now returns a dict with "annotations", "raw_sequences", "sequences", "modifications" and "ignored_lines"
  - Add get_mods_composition() that returns a composition from a list of Unimod PTMs
  - Remove the appended "\_stfi" if -o flag is provided

**0.3.0** (2019-04-18)

*Changed*
- Add support for files with annotations before the sequences
  - **Breaking change:** sequence_parser() now returns (annotations, sequences, ignored_lines)
  - seq_to_tsv() now accepts (sequences, unlabelled_aa, annotations=None)

*Fixed*
- Output files now have "\_stfi" appended to differentiate from .tsv input files with the same name

**0.2.1** (2019-04-17)
- Format CHANGELOG

**0.2.0** (2019-04-17)
- Add bumpversion

*Changed*
- seq_to_tsv() no longer writes a file, instead it returns a dataframe

**0.1.0** (2019-04-08)
- First release


tests/test_input.py

"""Test files provided."""

import seq_to_first_iso as stfi


def test_constants():
    assert stfi.AMINO_ACIDS == set("ACDEFGHIKLMNPQRSTVWY")
    assert stfi.isotopic_abundance ==  {"H[1]": 0.999885,
                                        "H[2]": 0.000115,
                                        "C[12]": 0.9893,
                                        "C[13]": 0.0107,
                                        "X[12]": 0.9893,
                                        "X[13]": 0.0107,
                                        "N[14]": 0.99632,
                                        "N[15]": 0.00368,
                                        "O[16]": 0.99757,
                                        "O[17]": 0.00038,
                                        "O[18]": 0.00205,
                                        "S[32]": 0.9493,
                                        "S[33]": 0.0076,
                                        "S[34]": 0.0429}




tests/test_seq_to_first_iso.py

"""Test files provided."""

import filecmp
from pathlib import Path

import pandas as pd
from pyteomics import mass
from pyteomics.auxiliary.structures import PyteomicsError
import pytest

import seq_to_first_iso as stfi

data_dir = Path(__file__).parent.resolve().joinpath("test_data")
REL = 1e-6


# Tests for pyteomics.
@pytest.fixture(scope="session")
def get_composition():
    return [("ACDE", {'H': 24, 'C': 15, 'O': 9, 'N': 4, 'S': 1}),
            ("VPKER", {'H': 49, 'C': 27, 'O': 8, 'N': 9}),
            ("ACDEFGHIKLMNPQRSTVWY", {'H': 159, 'C': 107, 'O': 30, 'N': 29, 'S': 2}),
            ("", {}),
            ("ACTG", {'H': 22, 'C': 12, 'O': 6, 'N': 4, 'S': 1}),
            ("LILI", {'H': 46, 'C': 24, 'O': 5, 'N': 4}),
            ("LLLL", {'H': 46, 'C': 24, 'O': 5, 'N': 4})
            ]


def test_composition(get_composition):
    assert mass.Composition("ACDE")
    assert mass.Composition("A") + mass.Composition("C")
    assert mass.Composition(parsed_sequence="ACDE") == {'H': 22, 'C': 15, 'O': 8, 'N': 4, 'S': 1}

    for data in get_composition:
        sequence = data[0]
        expected = data[1]
        assert mass.Composition(sequence) == expected


@pytest.fixture(scope="session")
def get_mass():
    return [("", 0),
            ("ACDE", 436.12639936491996),
            ("VPKER", 627.37040957111),
            ("ACDEFGHIKLMNPQRSTVWY", 2394.1249068251295),
            ("ACTG", 350.1260054421),
            ("LILI", 470.34682059221996),
            ("LLLL", 470.34682059221996),
            ("AAAAAALAAAAAIAAAAAAAAAAAAAGAAAAAAAAAAVIYWWSSEPED", 4108.065244683701),
            ]


def test_calculate_mass(get_mass):
    assert mass.calculate_mass("ACDE") == pytest.approx(436.12639936, REL)
    assert mass.calculate_mass(mass.Composition("ACDE")) == pytest.approx(436.12639936, REL)
    assert mass.calculate_mass(parsed_sequence="ACDE") == pytest.approx(418.115834, REL)
    assert mass.calculate_mass("A") == pytest.approx(89.04767846841, REL)

    for data in get_mass:
        sequence = data[0]
        expected = data[1]
        assert mass.calculate_mass(sequence) == pytest.approx(expected, REL)


# def test_fast_mass(get_mass):
#    assert mass.fast_mass("A")
#    assert mass.fast_mass2("A")
#
#    # fast_mass has a problem with empty sequences.
#    for data in get_mass[1:]:
#        sequence = data[0]
#        expected = data[1]
#        assert mass.fast_mass(sequence) == pytest.approx(expected, REL)
#        assert mass.fast_mass2(sequence) == pytest.approx(expected, REL)


# Tests for seq_to_first_iso.
def test_constants():
    assert stfi.AMINO_ACIDS == set("ACDEFGHIKLMNPQRSTVWY")
    assert stfi.NATURAL_ABUNDANCE == {"H[1]": 0.999885, "H[2]": 0.000115,
                                       "C[12]": 0.9893, "C[13]": 0.0107,
                                       "X[12]": 0.9893, "X[13]": 0.0107,
                                       "N[14]": 0.99632, "N[15]": 0.00368,
                                       "O[16]": 0.99757, "O[17]": 0.00038, "O[18]": 0.00205,
                                       "S[32]": 0.9493, "S[33]": 0.0076, "S[34]": 0.0429}

    assert stfi.C12_ABUNDANCE["C[12]"] == pytest.approx(0.9999, rel=REL)
    assert stfi.C12_ABUNDANCE["C[13]"] == pytest.approx(0.0001, rel=REL)
    assert {k: v for k, v in stfi.C12_ABUNDANCE.items()
            if k not in ["C[12]", "C[13]"]} == {"H[1]": 0.999885, "H[2]": 0.000115,
                                                "X[12]": 0.9893, "X[13]": 0.0107,
                                                "N[14]": 0.99632, "N[15]": 0.00368,
                                                "O[16]": 0.99757, "O[17]": 0.00038, "O[18]": 0.00205,
                                                "S[32]": 0.9493, "S[33]": 0.0076, "S[34]": 0.0429}


@pytest.fixture(scope="session")
def get_separated_sequence():
    """For C,D and W unlabelled."""
    return [("ABCDEFG", ("ABEFG", "CD")),
            ("ABCDEFGCDCDDC", ("ABEFG", "CDCDCDDC")),
            ("", ("", "")),
            ]


def test_separation(get_separated_sequence):
    test_unlabelled = "CDW"
    for data in get_separated_sequence:
        sequence = data[0]
        expected = data[1]
        assert stfi.separate_labelled(sequence, test_unlabelled)
        result = stfi.separate_labelled(sequence, test_unlabelled)
        assert type(result) is tuple
        assert result == expected


def test_parse_input_file(caplog):
    test_file = data_dir.joinpath("sample_sequence.tsv")
    parser_output = stfi.parse_input_file(test_file, sep="")
    assert "Separator is empty" in caplog.text
    parser_output = stfi.parse_input_file(test_file)
    assert type(parser_output) is pd.DataFrame
    assert parser_output.shape == (8, 3)

    with pytest.raises(FileNotFoundError):
        stfi.parse_input_file("not_a_file")

    empty_file = data_dir.joinpath("sample_empty_file.tsv")
    with pytest.raises(Exception):
        stfi.parse_input_file(empty_file)


def test_filter_input_dataframe(caplog):
    test_file = data_dir.joinpath("sample_sequence.tsv")
    parser_output = stfi.parse_input_file(test_file)
    filter_output = stfi.filter_input_dataframe(parser_output, "sequence", "charge")
    assert type(filter_output) is pd.DataFrame
    assert filter_output.shape == (8, 2)

    bad_file = data_dir.joinpath("bad_sample_sequence.txt")
    with pytest.raises(ValueError):
        parser_output = stfi.parse_input_file(bad_file)
        stfi.parse_input_file(parser_output)


@pytest.fixture(scope="session")
def get_amino_acid_sequences():
    return [("PEPTIDE", ("PEPTIDE", "")),
            ("ACDE", ("ACDE", "")),
            ("VPKER", ("VPKER", "")),
            ("ACDEFGHIKLMNPQRSTVWY", ("ACDEFGHIKLMNPQRSTVWY", "")),
            ("ACTGZ", ("", "Unrecognized amino acids.")),
            ("LILI", ("LILI","")),
            ("AAaag", ("", "Unrecognized amino acids.")),
            ("JOJO", ("", "Unrecognized amino acids.")),
            ]


def test_check_amino_acids(get_amino_acid_sequences):
    for data in get_amino_acid_sequences:
        sequence_input, result = data
        assert stfi.check_amino_acids(sequence_input) == result


def test_deprecated_computation_isotopologue():
    test_composition = mass.Composition("ACDE")
    stfi.compute_M0 = stfi.seq_to_first_iso.compute_M0
    stfi.compute_M1 = stfi.seq_to_first_iso.compute_M1
    assert stfi.compute_M0(test_composition, stfi.NATURAL_ABUNDANCE) == pytest.approx(0.77662382, REL)
    assert stfi.compute_M0(test_composition, stfi.C12_ABUNDANCE) == pytest.approx(0.911253268, REL)

    assert stfi.compute_M1(test_composition, stfi.NATURAL_ABUNDANCE) == pytest.approx(0.1484942353, REL)
    assert stfi.compute_M1(test_composition, stfi.C12_ABUNDANCE) == pytest.approx(0.0277650369575, REL)


def test_computation_isotopologue():
    # Standard formula.
    test_composition = mass.Composition("ACDE")
    assert stfi.compute_M0_nl(test_composition, stfi.NATURAL_ABUNDANCE) == pytest.approx(0.77662382, REL)
    assert stfi.compute_M0_nl(test_composition, stfi.C12_ABUNDANCE) == pytest.approx(0.911253268, REL)
    assert stfi.compute_M1_nl(test_composition, stfi.NATURAL_ABUNDANCE) == pytest.approx(0.1484942353, REL)
    assert stfi.compute_M1_nl(test_composition, stfi.C12_ABUNDANCE) == pytest.approx(0.0277650369575, REL)


def test_string_casting():
    assert stfi.formula_to_str(mass.Composition("ACDE")) == "C15H24O9N4S1"
    assert stfi.formula_to_str(mass.Composition("PEPTIDE")) == "C34H53O15N7"
    assert stfi.formula_to_str(mass.Composition("ACDEFGHIKLMNPQRSTVWY")) == "C107H159O30N29S2"


def test_convert_atom_C_to_X():
    assert stfi.convert_atom_C_to_X("ACDE") == mass.Composition({'H': 24, 'O': 9, 'N': 4, 'S': 1, 'X': 15})
    assert stfi.convert_atom_C_to_X("PEPTIDE") == mass.Composition({'H': 53, 'O': 15, 'N': 7, 'X': 34})
    assert stfi.convert_atom_C_to_X("ACDEFGHIKLMNPQRSTVWY") == mass.Composition({'H': 159, 'O': 30, 'N': 29, 'S': 2, 'X': 107})


@pytest.fixture(scope="session")
def get_charges():
    return [(1, mass.Composition({"H":1})),
            (2, mass.Composition({"H":2})),
            (3, mass.Composition({"H":3})),
            (4, mass.Composition({"H":4})),
            (0, mass.Composition({"H":0})),
            ]

def test_get_charge_composition(get_charges):
    get_charge_composition = stfi.seq_to_first_iso.get_charge_composition
    for charge, target_composition in get_charges:
        assert get_charge_composition(charge) == target_composition


@pytest.fixture(scope="session")
def get_mods():
    return [(["Oxidation"], mass.Composition({"O":1})),
            (["Acetyl", "Phospho"], mass.Composition({'H':3, 'C':2, 'O':4, "P":1})),
            (["Acetyl", "Phospho", "not_mod"], mass.Composition({'H':3, 'C':2, 'O':4, "P":1})),
            ([], mass.Composition()),
            ]


def test_get_mods_composition(get_mods, caplog):
    get_mods_composition = stfi.seq_to_first_iso.get_mods_composition
    get_mods_composition(["not_mod"])
    assert "entry not found" in caplog.text
    # Mod with element not in CHONPSX.
    get_mods_composition(["Heme"])
    assert "(Heme) is not supported" in caplog.text
    for data in get_mods:
        modifications = data[0]
        expected = data[1]
        assert get_mods_composition(modifications) == expected


def test_cli_parser(caplog):
    user_parser = stfi.seq_to_first_iso.user_input
    test_file = data_dir.joinpath("sample_sequence.tsv")

    assert user_parser([str(test_file), "sequence", "charge"])
    with pytest.raises(SystemExit):
        user_parser(["not_a_file", "sequence", "charge"])

    # It seems like a space between -u and A here won't be escaped.
    cli_labelled = user_parser([str(test_file), "sequence", "charge", "-uA,c"])
    assert cli_labelled.unlabelled_aa == ["A", "C"]


def test_main(caplog):
    test_file = data_dir.joinpath("sample_sequence.tsv")

    stfi.seq_to_first_iso.cli([str(test_file), "sequence", "charge", "-uA"])
    assert "Amino acid with default abundance" in caplog.text
    assert Path("sample_sequence_stfi.tsv").is_file()
    assert filecmp.cmp("sample_sequence_stfi.tsv", data_dir.joinpath("sample_sequence_stfi_ref_A.tsv"), shallow=False)


seq_to_first_iso/__init__.py
"""Compute intensities of the first two isotopologue.

Use peptide sequences and charges.

The program computes M0 and M1 and differentiate labelled
(with a 99.99 % C[12] enrichment) and unlabelled amino acids.

Read a .tsv file composed of amino acid sequences on each line and return:
sequence, mass, formula, formula_X, M0_NC, M1_NC, M0_12C and M1_12C
in a .tsv file.

Formula_X is the chemical formula with carbon of unlabelled
amino acids marked as X.

NC means Normal Condition, 12C means C[12] enrichment condition.


Example
-------
Running the script after installation

    $ seq-to-first-iso sequences.tsv sequence_column_name charge_column_name

will provide file 'sequences_stfi.tsv'


Notes
-----
Carbon of unlabelled amino acids keep default isotopic abundance,
and are represented as X in formulas.
Naming conventions for isotopes follow pyteomics's conventions.

"""

__authors__ = "Lilian Yang-crosson, Pierre Poulain"
__license__ = "BSD 3-Clause License"
__version__ = "1.1.0"
__maintainer__ = "Pierre Poulain"
__email__ = "pierre.poulain@cupnet.net"

from .seq_to_first_iso import (AMINO_ACIDS,
                               XTANDEM_MOD_PATTERN,
                               UNIMOD_MODS,
                               NATURAL_ABUNDANCE,
                               C12_ABUNDANCE,
                               parse_input_file,
                               filter_input_dataframe,
                               check_amino_acids,
                               separate_labelled,
                               compute_M0_nl,
                               compute_M1_nl,
                               formula_to_str,
                               convert_atom_C_to_X,
                               get_charge_composition,
                               get_mods_composition,
                               compute_intensities,
                               )

__all__ = ["AMINO_ACIDS",
           "XTANDEM_MOD_PATTERN",
           "UNIMOD_MODS",
           "NATURAL_ABUNDANCE",
           "C12_ABUNDANCE",
           "parse_input_file",
           "filter_input_dataframe",
           "check_amino_acids",
           "separate_labelled",
           "compute_M0_nl",
           "compute_M1_nl",
           "formula_to_str",
           "convert_atom_C_to_X",
           "get_charge_composition",
           "get_mods_composition",
           "compute_intensities",
           ]


seq-to-first-iso.py

"""Create a tsv file from a file of sequences.

Read a file composed of sequences of amino acids on each line and
return sequence, mass, formula, M0 and M1 in normal and C[12] conditions
as a tsv file.

Naming conventions for isotopes follow pyteomics's conventions.
"""

__authors__ = "Lilian Yang-crosson"
__copyright__ = ""
__credits__ = [""]
__license__ = ""
__version__ = "0.1"
__maintainer__ = ""
__email__ = ""

import argparse
from pathlib import Path

import pandas as pd
from pyteomics import mass


USAGE_ERROR = "Usage: python seq-to-first-iso.py filename " \
             + "[-o output] [-n aa]"
# Note: pyteomics also have H- and -OH that can be used for sequences
# which are not supported in this version.  They are implicitly added.
AMINO_ACIDS = set("ACDEFGHIKLMNPQRSTVWYUO")


def user_input():
    """
    Handle the user parameter from the command line.

    Parse the command line parameter and build the list of input files.
    """
    parser = argparse.ArgumentParser(
        description="Read a file of sequences and creates a tsv file")

    # Input file is required as a positional argument.
    parser.add_argument("input", type=Path, help="file to parse")

    # Optional arguments.
    parser.add_argument("-o", "--output", type=Path,
                        help="name of output file")
    parser.add_argument("-n", "--non_labelled_aa",
                        metavar="amino_a",
                        help="amino acids with default abundance")

    options = parser.parse_args()

    # Check if file exists.
    if not options.input.is_file():
        exit("Error: file {} does not exist in current directory {}\n".format(
            options.input, options.input.cwd()) + USAGE_ERROR)

    # Check if amino acids are correct.  If not, tell which one.
    if not options.non_labelled_aa:
        # Change to empty list to avoid Nonetype errors.
        options.non_labelled_aa = []
    else:
        options.non_labelled_aa = options.non_labelled_aa.split(",")
        # Convert amino acids to uppercase for compatibility.
        options.non_labelled_aa = [char.upper()
                                   for char in options.non_labelled_aa]
        unrecognized_aa = []

        for arg in options.non_labelled_aa:
            if arg not in AMINO_ACIDS:
                unrecognized_aa.append(arg)

        if unrecognized_aa:
            print("Warning: {} not recognized "
                  "as amino acid".format(unrecognized_aa))

    return options


def sequence_parser(file):
    """Return a list of peptide sequences parsed from a file.

    Take a Path() object as argument, return a list of uppercase peptides
    or exit if the list is empty.
    """
    # Obtain a list of sequences as string if they are amino acids.
    with open(file, "r") as filin:
        sequences = []
        ignored_lines = 0
        for sequence in filin:
            upper_sequence = sequence.upper().strip()
            # Character not recognized as amino acid.
            if not (set(upper_sequence) - AMINO_ACIDS) and upper_sequence:
                sequences.append(upper_sequence)
            else:
                ignored_lines += 1

    # Check if the file format is correct.
    if not sequences:
        exit("Error: incorrect format, make sure that lines "
             "in {} are valid sequences of amino acids".format(file))
    if ignored_lines:
        print("{} lines ignored out of {}".format(ignored_lines,
                                                  ignored_lines+len(sequences)
                                                  ))
    return sequences


def compute_M0(f, a):
    """Return the monoisotopic abundance M0 of a sequence with its formula.

    f is the chemical formula, as a dict of counts for each element:
        {element_name: count_of_element_in_sequence, ...}
    a is the abundance of isotopes, as a dict in the format:
        {element_name[isotope_number]: relative abundance, ..}

    """
    M0 = a["C[12]"]**f["C"] * a["H[1]"]**f["H"] * a["N[14]"]**f["N"] \
        * a["O[16]"]**f["O"] * a["S[32]"]**f["S"]
    return M0


def compute_M1(f, a):
    """Compute abundance of second isotopologue M1 from its formula.

    f is the chemical formula, as a dict of counts for each element:
        {element_name: count_of_element_in_sequence, ...}
    a is the abundance of isotopes, as a dict in the format:
        {element_name[isotope_number]: relative abundance, ..}

    """
    M1 = (
          (f["C"] * a["C[12]"]**(f["C"]-1) * a["C[13]"]
              * a["H[1]"]**f["H"]
              * a["N[14]"]**f["N"]
              * a["O[16]"]**f["O"]
              * a["S[32]"]**f["S"])

          + (f["H"] * a["C[12]"]**f["C"]
              * a["H[1]"]**(f["H"]-1) * a["H[2]"]
              * a["N[14]"]**f["N"]
              * a["O[16]"]**f["O"]
              * a["S[32]"]**f["S"])

          + (f["N"] * a["C[12]"]**f["C"]
              * a["H[1]"]**f["H"]
              * a["N[14]"]**(f["N"]-1) * a["N[15]"]
              * a["O[16]"]**f["O"]
              * a["S[32]"]**f["S"])

          + (f["O"] * a["C[12]"]**f["C"]
              * a["H[1]"]**f["H"]
              * a["N[14]"]**f["N"]
              * a["O[16]"]**(f["O"]-1) * a["O[17]"]
              * a["S[32]"]**f["S"])

          + (f["S"] * a["C[12]"]**f["C"]
              * a["H[1]"]**f["H"]
              * a["N[14]"]**f["N"]
              * a["O[16]"]**f["O"]
              * a["S[32]"]**(f["S"]-1) * a["S[33]"])
          )
    return M1


def separate_labelled(sequence, unlabelled_aa):
    """Get the sequence of unlabelled amino acids from a sequence.

    sequence is a string of amino acids
    unlabelled_aa is a container (list, string...) of unlabelled amino acids

    Return the sequences as a tuple of string with:
        the sequence without the unlabelled amino acids
        the unlabelled amino acids in the sequence

    Note: we can also use comprehension lists (might go faster)
    labelled_seq = "".join([char in sequence if char not in unlabelled_aa])
    unlabelled_seq = "".join([char in sequence if char in unlabelled_aa])
    """
    labelled_seq = []
    unlabelled_seq = []
    for char in sequence:
        if char in unlabelled_aa:
            unlabelled_seq.append(char)
        else:
            labelled_seq.append(char)
    return "".join(labelled_seq), "".join(unlabelled_seq)


def compute_M0_nl(f, a):
    """Return the monoisotopic abundance M0 of a formula with mixed labels.

    f is the chemical formula, as a dict of counts for each element:
        {element_name: count_of_element_in_sequence, ...}
    a is the abundance of isotopes, as a dict in the format:
        {element_name[isotope_number]: relative abundance, ..}

    X represents C with default isotopic abundance.
    """
    M0 = a["C[12]"]**f["C"] * a["X[12]"]**f["X"] * a["H[1]"]**f["H"] \
        * a["N[14]"]**f["N"] * a["O[16]"]**f["O"] * a["S[32]"]**f["S"]
    return M0


def compute_M1_nl(f, a):
    """Compute abundance of second isotopologue M1 from its formula.

    f is the chemical formula, as a dict of counts for each element:
        {element_name: count_of_element_in_sequence, ...}
    a is the abundance of isotopes, as a dict in the format:
        {element_name[isotope_number]: relative abundance, ..}

    X represents C with default isotopic abundance.
    """
    M1 = (
          (f["C"] * a["C[12]"]**(f["C"]-1) * a["C[13]"]
              * a["X[12]"]**f["X"]
              * a["H[1]"]**f["H"]
              * a["N[14]"]**f["N"]
              * a["O[16]"]**f["O"]
              * a["S[32]"]**f["S"])

          + (f["X"] * a["C[12]"]**f["C"]
              * a["X[12]"]**(f["X"]-1) * a["X[13]"]
              * a["H[1]"]**f["H"]
              * a["N[14]"]**f["N"]
              * a["O[16]"]**f["O"]
              * a["S[32]"]**f["S"])

          + (f["H"] * a["C[12]"]**f["C"]
              * a["X[12]"]**f["X"]
              * a["H[1]"]**(f["H"]-1) * a["H[2]"]
              * a["N[14]"]**f["N"]
              * a["O[16]"]**f["O"]
              * a["S[32]"]**f["S"])

          + (f["N"] * a["C[12]"]**f["C"]
              * a["X[12]"]**f["X"]
              * a["H[1]"]**f["H"]
              * a["N[14]"]**(f["N"]-1) * a["N[15]"]
              * a["O[16]"]**f["O"]
              * a["S[32]"]**f["S"])

          + (f["O"] * a["C[12]"]**f["C"]
              * a["X[12]"]**f["X"]
              * a["H[1]"]**f["H"]
              * a["N[14]"]**f["N"]
              * a["O[16]"]**(f["O"]-1) * a["O[17]"]
              * a["S[32]"]**f["S"])

          + (f["S"] * a["C[12]"]**f["C"]
              * a["X[12]"]**f["X"]
              * a["H[1]"]**f["H"]
              * a["N[14]"]**f["N"]
              * a["O[16]"]**f["O"]
              * a["S[32]"]**(f["S"]-1) * a["S[33]"])
          )

    return M1


def formula_to_str(composition):
    """Return formula from Composition as a string."""
    formula_str = ""
    for element in "CHONSX":
        if element in composition:
            formula_str += f"{element}{composition[element]}"
    return formula_str


def seq_to_midas(sequence_l, sequence_nl):  # ADDED
    """Take 2 amino acid sequences and return the formula for MIDAs.

    Note: the function assumes the second sequence has no terminii.
    """
    formula_l = mass.Composition(sequence_l)
    formula_nl = mass.Composition(parsed_sequence=sequence_nl)
    try:
        formula_nl["X"] = formula_nl.pop("C")
    except KeyError:
        pass
    return formula_l+formula_nl


def formula_to_midas(formula_l, formula_nl):
    """Take 2 chemical formulas and return the formula for MIDAs.

    Note: can adapt to take sequence to composition.
    """
    f2 = mass.Composition(formula_nl)  # to not change formula2
    try:
        f2["X"] = f2.pop("C")  # Based on the assumption there are C.
    except KeyError:
        pass
    return formula_l+f2


if __name__ == "__main__":
    # Default isotopic abundances from MIDAs website:
    # https://www.ncbi.nlm.nih.gov/CBBresearch/Yu/midas/index.html .
    # X is C with default abundance
    isotopic_abundance = {"H[1]": 0.999885, "H[2]": 0.000115,
                          "C[12]": 0.9893,  "C[13]": 0.0107,
                          "X[12]": 0.9893,  "X[13]": 0.0107,
                          "N[14]": 0.99632, "N[15]": 0.00368,
                          "O[16]": 0.99757, "O[17]": 0.00038, "O[18]": 0.00205,
                          "S[32]": 0.9493,  "S[33]": 0.0076,  "S[34]": 0.0429}

    C12_abundance = dict(isotopic_abundance)
    prop = 0.9999
    C12_abundance["C[12]"] = prop
    C12_abundance["C[13]"] = 1-prop

    options = user_input()
    input_file = options.input
    unlabelled_aa = options.non_labelled_aa

    if unlabelled_aa:
        print("Amino acid with default abundance: {}".format(unlabelled_aa))

    print("Parsing file")
    sequences = sequence_parser(input_file)

    # Dataframe of sequences.
    df_peptides = pd.DataFrame({"sequence": sequences})

    # Separate sequences.
    df_peptides["labelled"], df_peptides["unlabelled"] = zip(
            *df_peptides["sequence"].apply(separate_labelled,
                                           unlabelled_aa=unlabelled_aa))

    # Add mass and formulas of sequences
    print("Computing mass")
    df_peptides["mass"] = df_peptides["sequence"].map(mass.calculate_mass)

    print("Computing formula")
    # Formula as a string (instead of mass.Composition).
    df_peptides["f"] = df_peptides["sequence"].apply(mass.Composition)
    df_peptides["formula"] = df_peptides["f"].apply(formula_to_str)
    # Composition, with unlabelled C as element X.
    df_peptides["f_X"] = df_peptides.apply(lambda x:
                                           seq_to_midas(x["labelled"],
                                                        x["unlabelled"]),
                                           axis=1)

    # Add M0 and M1 in normal conditions.
    print("Computing M0 and M1")
    # Can use compute_M0_nl with isotopic abundance twice
    df_peptides["M0_NC"] = df_peptides["f_X"].apply(compute_M0_nl,
                                                    a=isotopic_abundance)
    df_peptides["M1_NC"] = df_peptides["f_X"].apply(compute_M1_nl,
                                                    a=isotopic_abundance)

    df_peptides["M0_12C"] = df_peptides["f_X"].apply(compute_M0_nl,
                                                     a=C12_abundance)
    df_peptides["M1_12C"] = df_peptides["f_X"].apply(compute_M1_nl,
                                                     a=C12_abundance)

    # For verification with MIDAs, might be removed.
    df_peptides["formula_X"] = df_peptides["f_X"].apply(formula_to_str)

    # Choose output filename.
    if not options.output:
        output_file = input_file.stem + ".tsv"
    else:
        output_file = options.output

    wanted_columns = ["sequence", "mass", "formula", "formula_X",
                      "M0_NC", "M1_NC", "M0_12C", "M1_12C"]
    # Import dataframe to tsv file.
    df_peptides[wanted_columns].to_csv(output_file, sep="\t", index=False)
